### PR TITLE
Create VM - BasicSettings, select automatically single option menu

### PR DESCRIPTION
**Pull request by bamsalem**
Fixes: #1252
**Depends on**: https://github.com/oVirt/ovirt-web-ui/pull/1291 (MERGED)

In the CreateVMWizard dialog -> "Basic settings" step the user had to select an option from list menu even if there is only one option allowed to the user.
It could be annoying and unnecessary, in this PR lists with single option will be automatically selected.

There are few changes which not related to the auto select and used to avoid the dialog close confirmation if the user not changed anything and the only changes were done are from the "auto select" feature, those changes made to maintain the behavior which created in #1231 .

### CODE REVIEW

**Comment by ovirt-infra** [comment]
Hello contributor, thanks for submitting a PR for this project! 

I am the bot who triggers "standard-CI" builds for this project.
As a security measure, I will not run automated tests on PRs that are not from white-listed contributors.

In order to allow automated tests to run, please ask one of the project maintainers to review the code and then do one of the following:

1. Type `ci test please` on this PR to trigger automated tests for it.
2. Type `ci add to whitelist` on this PR to trigger automated tests for it and also add you to the contributor white-list so that your future PRs will be tested automatically. ( keep in mind this list might be overwritten if the job XML is refreshed, for permanent whitelisting, please follow #3 option )
3. If you are planning to contribute to more than one project, maybe it's better to ask them to add you to the project organization, so you'll be able to run tests for all the organization's projects.

**Review by hstastna — CHANGES_REQUESTED** [review]
I've tested the changes with only one data center, one cluster, one template available. The available cluster was automatically selected in the drop down as expected, but after I changed _Provision Source_ to _Template_, the only one available template was not automatically selected. Instead, the _--Select Template --_ was displayed in the drop down:
select

**Comment by bamsalem** [comment]
Hi @hstastna,
I made few improvements.
the commits are separated to make it easier to understand what I did there.
I will squash the commits to one commit before moving to QE, so ignore that please, thanks

**Review by rszwajko — APPROVED** [review]
Well done!

**Comment by sjd78** [comment]
ci test

**Review by hstastna — DISMISSED** [review]
LGTM, selecting a single option automatically works well.

**Review by hstastna — COMMENTED** [review]
I had to review this PR again. Selecting a single option automatically works well in the Basic settings step. However I am still not sure about the refactoring part of this PR. It requires to test more and unrelated scenarios in the Basic settings step (unrelated to selecting single option automatically). We are mixing more stuff in this PR. And I am not sure if it makes the code more readable (IMHO the opposite, I think there weren't so many changes that we would have to move them to some new helper. So now the code for handling changes of Basic step is in more than a one place where it seems to me more complicated). Maybe we should have splitted that into two PRs.

@sjd78 please check this PR to move it forward :)

**Comment by michalskrivanek** [comment]
In general you should separate the actual fix and refactoring to separate PRs. We don't mix unrelated changes

**Comment by rszwajko** [comment]
> It requires to test more and unrelated scenarios in the Basic settings step

True. However every fix in this area would require such regression.

> We don't mix unrelated changes

Agree. However I don't see unrelated changes in this fix.

**Comment by bamsalem** [comment]
This PR depended on #1291 PR and should merge after it

File: src/components/CreateVmWizard/helpers.js
Comment (rszwajko) [inline src/components/CreateVmWizard/helpers.js:19]: Mutable `changes` object requires checking the whole method to determine what is returned.
For me it would be easer to return the complete object in each branch (and flatten the branch structure too)
i.e.
```
if (templateList.length > 2) {
   return {
        clusterId,
        dataCenterId,
        ...handleProvisionSourceChange(...)  
   }  
} else if (templateList.length == 2) {
....
} else if (templateList.length === 1 && templateList[0].id === blankTemplateId) {
  ...
  if (isoList.length === 1) {
     return { ...}
  else {
    return { ...}
  }
}
else {
 .....
}
```

**Review by rszwajko — CHANGES_REQUESTED** [review]

File: src/components/CreateVmWizard/steps/BasicSettings.js
Comment (bamsalem) [reply src/components/CreateVmWizard/steps/BasicSettings.js:197]: @rszwajko we chatted about that it done to silence eslint's error.

**Review by rszwajko — APPROVED** [review]
The described bug scenario works for me with this patch.

**Comment by sgratch** [comment]
@bamsalem it seems that there are conflicts between this PR and depended PR #1209:
both added src/components/CreateVmWizard/helpers.js with differences between the files.
both modified src/components/CreateVmWizard/steps/BasicSettings.js with differences between the files.

Need to solve all conflicts.

**Comment by bamsalem** [comment]
> @bamsalem it seems that there are conflicts between this PR and depended PR #1209:
> both added src/components/CreateVmWizard/helpers.js with differences between the files.
> both modified src/components/CreateVmWizard/steps/BasicSettings.js with differences between the files.
> 
> Need to solve all conflicts.

@sgratch 
I didn't have time to finish the rebase last week now it's re based.

**Review by sgratch — CHANGES_REQUESTED** [review]
Few comments
-  why the commit "refactor the changeHandlers from BasicSettings.js component to external helpers file" is duplicated both for this PR and also for https://github.com/oVirt/ovirt-web-ui/pull/1291? 

- In case of trying to create a VM in an environment with only once cluster but more than one template assigned to the default one cluster, there is a failure in this line: https://github.com/oVirt/ovirt-web-ui/blob/877dc37041a7b0063fe456f2137f96424db3fcb8/src/components/CreateVmWizard/steps/BasicSettings.js#L576 since sockets is not defined.

Check the stuck trace:
BasicSettings.js:575 Uncaught TypeError: Cannot read property 'sockets' of undefined
    at BasicSettings.render (BasicSettings.js:575)
    at finishClassComponent (react-dom.development.js:14741)
    at updateClassComponent (react-dom.development.js:14696)
    at beginWork (react-dom.development.js:15644)
    at performUnitOfWork (react-dom.development.js:19312)
    at workLoop (react-dom.development.js:19352)
    at HTMLUnknownElement.callCallback (react-dom.development.js:149)
    at Object.invokeGuardedCallbackDev (react-dom.development.js:199)
    at invokeGuardedCallback (react-dom.development.js:256)
    at replayUnitOfWork (react-dom.development.js:18578)

**Comment by bamsalem** [comment]
> * why the commit "refactor the changeHandlers from BasicSettings.js component to external helpers file" is duplicated both for this PR and also for #1291?

The commit is duplicated because this PR is based on #1291 and like Michal described in his [comment](https://github.com/oVirt/ovirt-web-ui/pull/1265#issuecomment-690296071) I separated them.

> * In case of trying to create a VM in an environment with only once cluster but more than one template assigned to the default one cluster, there is a failure in this line: https://github.com/oVirt/ovirt-web-ui/blob/877dc37041a7b0063fe456f2137f96424db3fcb8/src/components/CreateVmWizard/steps/BasicSettings.js#L576
>    since sockets is not defined.
> 
> Check the stuck trace:
> BasicSettings.js:575 Uncaught TypeError: Cannot read property 'sockets' of undefined
> at BasicSettings.render (BasicSettings.js:575)
> at finishClassComponent (react-dom.development.js:14741)
> at updateClassComponent (react-dom.development.js:14696)
> at beginWork (react-dom.development.js:15644)
> at performUnitOfWork (react-dom.development.js:19312)
> at workLoop (react-dom.development.js:19352)
> at HTMLUnknownElement.callCallback (react-dom.development.js:149)
> at Object.invokeGuardedCallbackDev (react-dom.development.js:199)
> at invokeGuardedCallback (react-dom.development.js:256)
> at replayUnitOfWork (react-dom.development.js:18578)

Regarding this I found the issue it was the default values from the CreateVmWizard.
I'm not sure which PR cause that but the cpu topology doesn't have default values so  I added that values in different commit thanks :)

**Comment by sgratch** [comment]
@bamsalem 
> The commit is duplicated because this PR is based on #1291 and like Michal described in his comment I separated them.

Right, so this PR fix can't compiled successfully without PR #1291 and PR #1291 should be merged first before this one (as written above as "Depends on: #1291" ). 
But we shouldn't duplicate the code on those 2 PRs because what's the point of the dependency if the code existed in this PR fix as well.... 

> Regarding this I found the issue it was the default values from the CreateVmWizard.
> I'm not sure which PR cause that but the cpu topology doesn't have default values so I added that values in different commit 

+1, Looks good now

**Comment by bamsalem** [comment]
I must use the code from #1291 I separated the PRs because Hilda and Michal asked that.
If it's unnecessary maybe we should close #1291 WDYT?

**Comment by sgratch** [comment]
> I must use the code from #1291 I separated the PRs because Hilda and Michal asked that.
> If it's unnecessary maybe we should close #1291 WDYT?

@bamsalem I think it still worth separating between extracting the handles which is a general refactoring and this PR which adds a new functionality. 
So based on the fact that those 2 PRs are depended, I suggest the following:
- merging PR #1291
- remove the duplicated code that exists already on the PR #1291
- re-basing this PR so that CI will pass successfully
- merging this PR once QE approves.

Sounds ok?

**Comment by sgratch** [comment]
@bamsalem the depended PR #1291 is merged, please remove duplicated code and rebase.

**Comment by bamsalem** [comment]
> @bamsalem the depended PR #1291 is merged, please remove duplicated code and rebase.

Done, I removed the duplicated code, and the PR is rebased

**Review by sgratch — APPROVED** [review]
LGTM

File: src/components/CreateVmWizard/helpers.js
Comment (hstastna) [inline src/components/CreateVmWizard/helpers.js:33]: Could you, please, change this comment a little bit? It's not 100% clear and gramatically correct, maybe you just wanted to write "cluster includes only one template (Blank)". Or maybe "cluster which includes only one template, has Blank template"?

File: src/components/CreateVmWizard/helpers.js
Comment (hstastna) [inline src/components/CreateVmWizard/helpers.js:47]: I would change this comment to "Cluster does not include any non blank template".

Regarding the "or iso cd" part of this comment, are you sure about that? Could you please explain? I don't see checking `isoList` here, only in the `if (templateList.length === 1 && templateList[0].id === blankTemplateId)` block.

**Review by hstastna — CHANGES_REQUESTED** [review]
I've added some little suggestions regarding improving some of code comments to make them more clear. And I've found some scenario in which selecting the single option automatically does not work properly and should be fixed:

0. Have only one Cluster, one Template - Blank available and no any ISO
1. Fill in the _Name_ in the Create VM Wizard
   (the only available Cluster and the Template are selected as expected)
2. Change the _Provision Source_ to _ISO_ 
3. Change the Provision Source back to _Template_
=> the only available Template is not selected automatically as expected, '-- Select Template --' is displayed instead:
select

**Comment by sgratch** [comment]
@hstastna 

> I've added some little suggestions regarding improving some of code comments to make them more clear. And I've found some scenario in which selecting the single option automatically does not work properly and should be fixed:
> 
>     Have only one Cluster, one Template - Blank available and no any ISO
>     Fill in the Name in the Create VM Wizard
>     (the only available Cluster and the Template are selected as expected)
>     Change the Provision Source to ISO
>     Change the Provision Source back to Template
>     => the only available Template is not selected automatically as expected, '-- Select Template --' is displayed instead:

This is part of the original requirements since we decided that once the user manually change the provision source field, the template/iso are not auto selected anymore. The auto selection is relevant only for 1st time the dialog is opened to make the user life easier. Once he started to manually change the selections then let him select the next fields as well..

File: src/components/CreateVmWizard/helpers.js
Comment (bamsalem) [reply src/components/CreateVmWizard/helpers.js]: @hstastna Sure, `isoList.length` range is from 0 to n (n = positive number greater than 0 | 1,2,etc...)
In line 36 we are checking if `isoList.length` value is greater than zero:
https://github.com/oVirt/ovirt-web-ui/blob/8c4c7e3fdbb6cd84890e7e2a7472fe070b28279f/src/components/CreateVmWizard/helpers.js#L36
If the result of that condition is `false` so we're jumping into the else block and it means that `isoList.length === 0`, in other words there is no iso for that cluster

File: src/components/CreateVmWizard/helpers.js
Comment (hstastna) [inline src/components/CreateVmWizard/helpers.js:47]: The cluster "includes" or "does not" include any non blank template here? Could you, please, change this comment to "cluster does not include any non blank template or iso cd"? Let's have the comments grammatically correct to prevent any potential misunderstandings ;) Thanks.

**Review by hstastna — CHANGES_REQUESTED** [review]
Except that I have a different point of view regarding the https://github.com/oVirt/ovirt-web-ui/pull/1265#issuecomment-729910754, LGTM. Just one more little thing to change and we are good to go 👍🏽

File: src/components/CreateVmWizard/helpers.js
Comment (bamsalem) [reply src/components/CreateVmWizard/helpers.js:47]: You are right, I wasn't focused, I missed that little 's', Thanks :)

Comment (sgratch) [reply src/components/CreateVmWizard/helpers.js]: > You are right, I wasn't focused, I missed that little 's', Thanks :)

looks ok now, so merging.

**Comment by isaranova** [comment]
LGTM

**Pull request merged**
